core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsEndpointGroup.java

```diff
@@ -0,0 +1,239 @@
+                     DnsServerAddressStreamProvider serverAddressStreamProvider,
+                     Backoff backoff, Iterable<DnsQuestion> questions,
+                     Consumer<DnsNameResolverBuilder> resolverConfigurator) {
+
+        this.eventLoop = eventLoop;
+        this.minTtl = minTtl;
+        this.maxTtl = maxTtl;
+        this.backoff = backoff;
+        this.questions = ImmutableList.copyOf(questions);
+        assert !this.questions.isEmpty();
+        logger = LoggerFactory.getLogger(getClass());
+        logPrefix = this.questions.stream()
+                                  .map(DnsQuestion::name)
+                                  .distinct()
+                                  .collect(Collectors.joining(", ", "[", "]"));
+
+        final DnsNameResolverBuilder resolverBuilder = new DnsNameResolverBuilder(eventLoop)
+                .channelType(TransportType.datagramChannelType(eventLoop.parent()))
+                .ttl(minTtl, maxTtl)
+                .nameServerProvider(serverAddressStreamProvider);
+
+        resolverConfigurator.accept(resolverBuilder);
+        resolver = resolverBuilder.build();
+    }
+
+    final Logger logger() {
+        return logger;
+    }
+
+    final String logPrefix() {
+        return logPrefix;
+    }
+
+    /**
+     * Invoke this method at the end of the subclass constructor to initiate the queries.
+     */
+    final void start() {
+        checkState(!started);
+        started = true;
+        eventLoop.execute(this::sendQueries);
+    }
+
+    private void sendQueries() {
+        if (stopped) {
+            return;
+        }
+
+        final Future<List<DnsRecord>> future;
+        final int numQuestions = questions.size();
+        if (numQuestions == 1) {
+            // Simple case of single query
+            final DnsQuestion question = questions.get(0);
+            logger.debug("{} Sending a DNS query", logPrefix);
+            future = resolver.resolveAll(question);
+        } else {
+            // Multiple queries
+            logger.debug("{} Sending DNS queries", logPrefix);
+            @SuppressWarnings("unchecked")
+            final Promise<List<DnsRecord>> aggregatedPromise = eventLoop.newPromise();
+            final FutureListener<List<DnsRecord>> listener = new FutureListener<List<DnsRecord>>() {
+                private final List<DnsRecord> records = new ArrayList<>();
+                private int remaining = numQuestions;
+
+                @Override
+                public void operationComplete(Future<List<DnsRecord>> future) throws Exception {
+                    if (future.isSuccess()) {
+                        final List<DnsRecord> records = future.getNow();
+                        if (aggregatedPromise.isDone()) {
+                            // parentFuture has been notified already, which means the previous query has
+                            // failed; discard the records since it's not useful anymore.
```

Comment:
Shouldn't we only fail if all queries fail?

Reply:
Hmm - I wonder if there's a case where one query will consistently fail for an extended period of time, in which case I guess the working records should still be used (it's why Java queries for both IPv4 and IPv6 even though it almost always only uses IPv4, but it could use IPv6 if that was failing for some reason)

Reply:
I see. Let me update this.